### ISSUE TITLE
Protocol Version should not be part of the connection info struct or URL

### ISSUE THREAD

**Issue by bikeshedder** [mitsuhiko/redis-rs#1447]
As I was working on redis PubSub for `deadpool-redis` I was wondering why the protocol version is part of the `ConnectionInfo` struct and `URL`. If I'm not mistaken changing the protocol version does require a code change. Therefore it isn't something that should ever be changed by a user.

Therefore I'd like to suggest the removal of this information from the `ConnectionInfo` struct and URL. This should be passed as a separate parameter instead. If a software depends on `RESP3` passing in a non-RESP3 URL will fail. Likewise a software expecting RESP2 responses will likely fail if a `RESP3` URL is passed in.

Maybe it would even make sense to make this two separate client structs so it's very clear that they're in essence deemed incompatible?

Related issues: 

- #1435
- bikeshedder/deadpool#226

**Comment by nihohit** [comment]
Hi, do you mind explaining why you don't think the protocol belongs in `RedisConnectionInfo`? Specifically, can you explain in what sense is it different from the other fields on the struct?

**Comment by bikeshedder** [comment]
`host`, `port`, and `database` are environment-specific and can be safely configured without affecting the application logic. The protocol version is different as it can cause breakage and require code changes.

As RESP3 is not backwards compatible it's not even an option to implement some kind of auto negotiation unless the redis client library implements some special logic to convert between the two protocol versions.

When you write an application that depends on RESP3 there is no turning back. If the user of such application does specify a RESP2 URL in the config it will just break. Also the user has to specify the RESP3 protocol version in the config which is even weirder.

A typical application would specify "Requires Redis/Valkey 6+". I don't see any use case where a user of such application would want to use an older/newer protocol version and break the application code by doing so.

**Comment by altanozlu** [comment]
Most of the stuff are compatible between RESP2&RESP3 in redis-rs, but I agree if an application is built on a feature requires RESP3 (e.g. MultiplexedConnection's subscribe feature) it won't work with an url uses RESP2(default). It can cascade to users has no idea about redis-rs when an application depends on redis-rs indirectly. Keeping this as part of redis uri makes that application let's opt in for RESP3 without an external configuration which is something positive IMHO.  Maybe before reaching [1.0](https://github.com/redis-rs/redis-rs/issues/1043), redis-rs should make RESP3 as default. Since EOL of redis versions that doesn't support RESP3 has already passed.


https://endoflife.date/redis
https://redis.io/docs/latest/operate/rs/installing-upgrading/product-lifecycle/

**Comment by nihohit** [comment]
Sorry, I didn't understand your explanation at all. Even if I agreed that setting `protocol` isn't backwards compatible and different from all other values*, why does that mean that it shouldn't be in the connection info? Where would you configure this, if not in the initial connection configuration?
And when you say it shouldn't be in the URL - well, have you checked the [Redis URL specification](https://github.com/redis/redis-specifications/blob/1252427cdbc497f66a7f8550c6b5f2f35367dc92/uri/redis.txt#L106)?

If your argument is that an application or a library that receives a URL from its users has to be able to handle all protocols, I think this is a wonderful example of input sanitization. If you can't handle a certain input value, sanitize  your inputs.

*I don't - if you play with `database` or `username` as part of the ACL, you still affect change which affects all services that touch this server, not only the current program.

**Comment by bikeshedder** [comment]
> And when you say it shouldn't be in the URL - well, have you checked the [Redis URL specification](https://github.com/redis/redis-specifications/blob/1252427cdbc497f66a7f8550c6b5f2f35367dc92/uri/redis.txt#L106)?

It's only a "Reserved URI field".

I consider it a mistake to mention this field in the spec. This creates the impression that it should be used to switch between protocol versions when implementing client libraries.

---

Consider the following statements:

1. An application that uses RESP3 features will fail when using a URL that lacks the `protocol` information
2. An application that is written with RESP2 in mind can crash or behave badly when using a `RESP3` URL.

The application **code** is the deciding factor if RESP2 or RESP3 can be used.

RESP3 is explicitly stated to be backwards incompatible to RESP2. If I'm not mistaken this is nothing the client library can fully abstract away and make transparent to the user of the client API.

This would be a non-issue if RESP3 was fully backwards compatible with RESP2.

---

> Sorry, I didn't understand your explanation at all. Even if I agreed that setting protocol isn't backwards compatible and different from all other values*, why does that mean that it shouldn't be in the connection info? Where would you configure this, if not in the initial connection configuration?

`RedisConnectionInfo` in itself is fine to have that `protocol` field. It only is problematic because of the URL parser which creates `(Redis)ConnectionInfo` structs with the `protocol` field filled.

As it stands right now an application that depends on RESP3 features can only parse the URL and override the version field to RESP3. Rendering the `protocol` info in the URL essentially useless:

```rust
let mut connection_info = env::var("REDIS_URL")?.into_connection_info()?;
connection_info.redis.protocol = ProtocolVersion::RESP3;
```

---

That's why I think the `protocol` should not be part of the URL and the `RedisConnectionInfo` struct.

## Proposed changes:

- Remove `protocol` from `RedisConnectionInfo` struct
- Remove `protocol` from URL parser
- Change the `Client::open` signature (and signature of the other methods) to include the protocol version as second parameter:
  ```rust
  pub fn open<T: IntoConnectionInfo>(params: T, protocol: ProtocolVersion) -> RedisResult<Client>
  ```

**Comment by bikeshedder** [comment]
## Alternative changes
- Leave `Client::open` as is
- Clean up the `get_*_connection_*` methods and replace them by a few methods like: `Client::get_async_resp3_connection(config: Resp3Config)`

Personally I've never understood why it's a two step process: `Client::open` + `Client::get_*_connection_*`. Is there any reason except for backwards compatibility why this has to be two method calls?

**Comment by nihohit** [comment]
> It only is problematic because of the URL parser which creates (Redis)ConnectionInfo structs with the protocol field filled.

I'm sorry, but it seems like there's a gap there. I agree that an app that is written to RESP2/3 could crash & burn when handling RESP3/2. From my experience this isn't true for the vast majority of apps, who only use variations of gets & sets, but theoretically - yes, the protocols aren't compatible.
What does URL parsing has to do with it? And why does this mean that we should break backwards compatibility with previous versions of the crate? Because all of your solutions are backwards incompatible, and I still don't see how your solutions solve the issue of protocol incompatibility.

> Is there any reason except for backwards compatibility why this has to be two method calls?

not AFAIK. IMO it's pure backwards compatibility.

**Comment by bikeshedder** [comment]
> Because all of your solutions are backwards incompatible, and I still don't see how your solutions solve the issue of protocol incompatibility.

It doesn't solve the incompatibility but lets the **code** decide what protocol version to use. Not the **user** who provides the configuration.

---

I just don't think the `version` should be part of any user provided configuration at all.

In my own code I always have to override the `version` part of the `ConnectionInfo` as I don't want to make this part of the application config.

Yes, removing this field is a backwards incompatible change. It's a good thing though as it improves safety and fixes some usability issues that I (and possibly others?) ran into.

The `redis` crate has had `27` backwards incompatible releases so far. Even if this is nothing to be considered for the next 0.28, 0.29,... release it's something that should be considered for 1.0.

**Comment by bikeshedder** [comment]
Sorry for being so annoying about this.
Feel free to close this issue if you disagree with what I said.

I'm going to remove the `protocol` from the `deadpool-redis` config structures and override the protocol version internally. I just wanted to share my thoughts here as I consider this an issue of the upstream `redis`-crate.

**Comment by nihohit** [comment]
> Sorry for being so annoying about this.

Not at all! There's always room for discussion, even if it takes some rounds to fully flesh out a discussion :)

> lets the code decide what protocol version to use. Not the user who provides the configuration.

Right, so this goes back to my earlier question. If I understand you correctly, the problematic scenario is with middleware (application or crate), where the user provides a URL to something that wraps `redis-rs`, but the user isn't working on the code using that configuration. As I said, in such a situation I would recommend sanitizing the inputs - in general, I don't think that middleware should directly expose dependency configurations as its own configuration, and a redis URL is exactly that.

> The redis crate has had 27 backwards incompatible releases so far.

True, but I'd like to think that the majority of those were for issues bigger than just usability - either incorrectness, or big footguns. Personally, while I agree that this is a footgun, I don't think it's a big one. 
That being said, we have an issue for things we'd like to break - https://github.com/redis-rs/redis-rs/issues/1043 . Feel free to suggest a concrete suggestion, with a link to the discussion here, there.

**Comment by bikeshedder** [comment]
By "user" I mean someone running an application that was built with the `redis` crate and not a developer building software with the crate.

Typically the application would read the `REDIS_URL` from the environment or some configuration file which is then passed into the `Client::open` function. Now the user is in charge of specifying the protocol version albeit the code written by the developer does require a specific protocol version to run reliably.

**Comment by bikeshedder** [comment]
@nihohit Please consider removing the protocol parameter from the URL prior the 1.0 release. The connection string really should not contain this kind of information. An application written with RESP3 in mind will just run into issues if configured with a RESP2 URL.

**Comment by nihohit** [comment]
I'm still willing to discuss this, but since this is backwards incompatible, I'd like to be sure that the need justifies the potential breakage.
ATM the user-provided string is converted into `ConnectionInfo` (probably when calling `Client::open`, but this can also be handled by the calling code) which contains the set protocol.
You say that an application relies on a certain protocol being set - how is it being set, if a URL is converted directly to `ConnectionInfo`? If it is being set on the created `ConnectionInfo`, then the protocol in the user provided URL is overwritten.

Essentially, the only problematic situation is one in which what you call the code part neither sanitizes the user provided URL, nor overwrites the protocol field on the created `ConnectionInfo` object, right? And since it can't provide the protocol in any other way, the problem you describe is relevant only for code which implicitly, not explicitly, relies on a certain protocol version being used.
Am I right in this analysis? Did I miss a scenario?

If I missed something, I'd appreciate a code snippet that demonstrates what I missed.

and just to be sure - when you say `a RESP2 URL`, you mean a URL that contains `protocol=resp2`, right? not a URL that doesn't specify a protocol, and so chooses the default?

**Comment by bikeshedder** [comment]
My thinking here is the following. Most users will use the `URL` in a configuration file or environment variable. Such as...

```rs
let redis_url = std::env::var("REDIS_URL")?;
```

...and use that directly to connect to the Redis database.

Let's assume the software requires RESP3 to work correctly. It's very tempting to write in the README:

> Please make sure to include `protocol=resp3` in the URL for the software to work correctly.

Of course I can just override the `protocol` after the URL is parsed. In the current state that's actually the correct way of doing it safely. And this brings me to my conclusion here. If overriding the `protocol` should be done in a software that depends on RESP3 it should not have been part of the URL in the first place.

The only situation I can think of where this could be useful is a software that supports both RESP2 and RESP3 and wants to give that power to the user to switch between those two. But I'm hard pressed to say that such software is either very niche or doesn't exist at all.

> and just to be sure - when you say a RESP2 URL, you mean a URL that contains protocol=resp2, right? not a URL that doesn't specify a protocol, and so chooses the default?

Yes.

**Comment by bikeshedder** [comment]
Just for the fun of it I asked ChatGPT to summarize the issue and take position: https://chatgpt.com/share/68f7463d-ea64-800e-b770-5a5561387e89

Prompt:
> Can you summarize https://github.com/redis-rs/redis-rs/issues/1447 for me and take position?

Answer:
> The GitHub issue #1447 in the `redis-rs` repository discusses the inclusion of the Redis protocol version in the connection information structure (`ConnectionInfo`) and URL. The author, @bikeshedder, argues that the protocol version is an internal implementation detail and should not be exposed to users. They suggest that the protocol version is unlikely to change dynamically and propose removing it from the connection info and URL, instead passing it as a separate parameter. This approach would clarify the distinction between the connection details and the protocol version, potentially leading to a more robust and user-friendly API.
>
> The discussion touches on the design of the Redis client and its flexibility. Some contributors express concerns that removing the protocol version from the connection info could complicate the client's ability to support multiple Redis versions simultaneously. However, the consensus leans towards simplifying the API by decoupling the protocol version from the connection details, thereby reducing complexity and potential for misuse.
> 
> **Position:**
> 
> I support the proposal to remove the protocol version from the connection info and URL. Exposing internal implementation details can lead to misuse and unnecessary complexity. By passing the protocol version as a separate parameter, the API becomes cleaner and more intuitive, aligning with best practices in software design that advocate for encapsulation and separation of concerns.
> 
> This change would also enhance the flexibility of the client, allowing it to adapt more easily to future Redis protocol versions without requiring users to modify their connection configurations. Therefore, I believe this adjustment would lead to a more robust and user-friendly Redis client in Rust.

I don't trust AI. I was just curious to see what GPT-5 "thinks" of this issue...

**Comment by nihohit** [comment]
> In the current state that's actually the correct way of doing it safely.

and after the change, it would be the only way of doing this. Essentially, code that requires the usage of RESP3 has to set the protocol explicitly, since it isn't the default. You're right that the code could ask the users to set the protocol in the connection string, but that's not as robust as overriding the field directly.
So unless I'm missing something, code that requires RESP3 isn't actually affected by this suggestion - either way, it has to parse the URL to `ConnectionInfo`, and then set the protocol on that connection info. Have I missed something?

**Comment by bikeshedder** [comment]
I would expect the `version` to be a parameter of the `open` call:
```diff
-let client = redis::Client::open("redis://127.0.0.1/?protocol=resp3").unwrap();
+let client = redis::Client::open("redis://127.0.0.1/", ProtocolVersion::RESP3").unwrap();
```

Alternatively there could be two open calls `open_resp2` and `open_resp3`:
```diff
-let client = redis::Client::open("redis://127.0.0.1/?protocol=resp3").unwrap();
+let client = redis::Client::open_resp3("redis://127.0.0.1/").unwrap
```

That `open` call could then be marked as `deprecated` and an alias to `open_resp2` to keep backwards compatibility.

**Comment by nihohit** [comment]
> I would expect the version to be a parameter of the open call:

That pattern is harder to maintain. As the library evolves, opening requires more parameters, which means that either we [accumulate more and more functions](https://github.com/redis-rs/redis-rs/pull/1672/files#diff-16703442692feae1ed03044b55412a59e037b4a1620d025361fc25fee4bffce8), or we break the existing function by adding more paratemers. Instead, `open` uses the idiomatic Rust pattern of using a configuration object, or any object easily convertible to it.

Now, since you haven't actually answered my question explicitly, I take from your answer that you agree with me - either way, the calling code needs to explicitly opt-in to RESP3. So, in such a situation, what's the problem with the calling code converting the input string to `ConnectionInfo`, and setting the protocol manually?

**Comment by bikeshedder** [comment]
Making the `protocol` part of the `ConnectionInfo` is fine. I just want to give my reasoning why I would prefer it to be outside of the `ConnectionInfo` struct:

In `deadpool-redis` I have created a 1:1 copy from it in order to support `serde::Deserialize` so it can be used to deserialize configuration files: https://docs.rs/deadpool-redis/latest/deadpool_redis/struct.ConnectionInfo.html

As it is a 1:1 copy the [RedisConnectionInfo](https://docs.rs/deadpool-redis/latest/deadpool_redis/struct.RedisConnectionInfo.html) also contains the `protocol` field. As written in deadpool-rs/deadpool#226 I'll be removing this field though as it is nothing that should be part of a configuration file.

Ideally the `redis` crate would implement `serde::Deserialize` for the `ConnectionInfo` and I can get rid of my 1:1 copy. That would however require that the `version` is not part of the struct.

**Comment by bikeshedder** [comment]
For `deadpool-redis` I'd be happy if `URL` just dropped support for specifying the protocol version. It is a popular way of specifying the connection to Redis. As long as I have to maintain a 1:1 copy of the `ConnectionInfo` (in order to add `serde` support) I can just skip that field in my own structs. I just want to avoid maintaining my own `parse_redis_url`.

I know I could just ignore the existence of that field and always override it. It just feels wrong to do so.

---

btw. if you really want to keep that field around and have it supported in the URL at least make the `protocol` field an `Option<ProtocolVersion>`:

```diff
 pub struct RedisConnectionInfo {
     pub db: i64,
     pub username: Option<String>,
     pub password: Option<String>,
-    pub protocol: ProtocolVersion,
+    pub protocol: Option<ProtocolVersion>,
 }
```

That way the absence of `?protocol=` in the URL will result in an `protocol=None` and I can return a ConfigurationError if the user tries to specify the protocol via the URL.

**Comment by bikeshedder** [comment]
btw. I just realized that `parse_redis_url` doesn't even raise an error if it contains any unsupported query parameters. A simple typo like `?procotol=resp3` will not result in an error. 🙈 

I never actually looked at that code and just assumed it would raise an error for unsupported query parameters.

**Comment by bikeshedder** [comment]
After reflecting over this for a bit...

I don't like that the protocol version can be specified via the URL but as the code doesn't return an error for unsupported query parameters there is little to be gained removing the support for it. Even though I think it's a huge mistake that the protocol version can be specified via the URL just removing it won't improve error reporting.

Users of the redis crate should always specify the protocol version and not rely on it being passed via the URL. Having a default for the protocol version in the `RedisConnectionInfo` struct is a mistake.

**Comment by nihohit** [comment]
> Having a default for the protocol version in the RedisConnectionInfo struct is a mistake.

Fair enough, I agree with that, and I think that's a backwards compatible change, too. woudl you like to open a PR for that?
